sklearn/cluster/_kmeans.py

```diff
@@ -1866,6 +2067,15 @@ def __init__(
     def _check_params_vs_input(self, X):
         super()._check_params_vs_input(X, default_n_init=3)
 
+        if self.tol > 0:
+            if sp.issparse(X):
+                variances = mean_variance_axis(X, axis=0)[1]
```

Comment:
Ditto: I find this is a bit more understandable: it makes explicit that `means` and `sum_weights` are ignored. 
```suggestion
            _, variances, _ = mean_variance_axis(X, axis=0)
```

What do you think?